Portability + Arc GPU fixes + concurrency v2 improvements

buggerman opened the pull request:
## Summary

Originally a small bc→awk portability change for TrueNAS, this PR has grown to cover a few related fork-driven fixes. Happy to split into smaller PRs if preferred — leaving as one for now since the script-level changes are tightly connected.

## Changes

### TrueNAS / minimal-system portability
Replace all `bc` usage with `awk`. `bc` is not installed by default on TrueNAS 25.10 and other minimal/appliance systems; `awk` is POSIX-standard and universally available. All `awk` replacements produce identical numeric results to the `bc` originals.

- `quicksync-benchmark.sh`: every `bc` call replaced (including the 4 new ones added by upstream's Benchmark Suite v2 in `run_concurrent_test` / `find_max_concurrency`)
- Dependency check now verifies `awk` instead of `bc`
- `README.md`, `web/src/pages/about/AboutPage.tsx`, `docs/k8s-plan.md` updated

### Arrow Lake power measurement reliability
Fixes discovered while testing Arrow Lake CPUs on TrueNAS where `intel_gpu_top` power data was missing or rejected:

- AWK division-by-zero guards in FPS / speed / power averages when ffmpeg or `intel_gpu_top` produces no valid samples — defaults to `0.00` instead of crashing
- `intel_gpu_top` version probe changed from `--version` to `-h` (older versions don't accept `--version`)
- Allow submission with rejected power readings (the backend already flags these via `arrow_lake_power_issue`); performance metrics remain useful even when wattage is unreliable

### FPS parsing fixes
- Match single-digit frame rates (the old `fps=.[1-9][1-9].` pattern excluded values < 10, common in 4K HEVC encodes)
- Match optional space after `=` (`fps= 52.37` format used by ffmpeg for slower encodes)

### Arc GPU codec detection
`check_qsv_encoder` was probing with `nullsrc=s=64x64:d=0.1`, which silently fails to initialize VP9/AV1 hardware on Arc A310 / Battlemage. Bumped to `640x480:d=1` so the probe survives encoder init. This is what made Arc users see VP9/AV1 reported as unsupported.

### Concurrency suite improvements
Acting on user feedback that the 10x cap is too low for Arc GPUs (A310 was hitting 2.18x at 10x for H.264 1080p with no way to find its actual ceiling):

- Default cap raised from 10 to 20. The existing `speed >= 1.0x` short-circuit still ends runs early on weaker hardware, so this costs nothing on the low end.
- `--max-concurrency N` flag for explicit override, validated as a positive integer
- Concurrency results header is now generated dynamically from the cap (the API parser in `parser.ts` already accepts variable-length speed columns)
- `--concurrency-only` flag skips the single-stream phase and runs only concurrency tests
- `--concurrency` keeps existing semantics (both phases) for backwards compatibility

### CLI ergonomics
- `--no-submit` flag as an alias for `QUICKSYNC_NO_SUBMIT=1` so users can opt out without exporting an env var

## Compatibility

- All shell changes are POSIX-compliant
- Output format unchanged for standard benchmark rows
- Concurrency rows now have variable column count (1x..Nx based on cap); already supported by the API parser
- No changes to API submission semantics

buggerman committed:
─────┬──────────────────────────────────────────────────────────────────────────
     │ STDIN
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Replace bc with awk for improved portability
   2 │
   3 │ Replace all bc usage with awk to enable running the benchmark on systems
   4 │ like TrueNAS 25.10 without installing additional packages. awk is a POSIX
   5 │ standard tool available by default on most Unix systems.
   6 │
   7 │ Changes:
   8 │ - Replace bc calculations with awk in quicksync-benchmark.sh
   9 │ - Update dependency check to verify awk instead of bc
  10 │ - Update README.md requirements and installation instructions
  11 │ - Update web/src/pages/about.astro requirements list
  12 │ - Update docs/k8s-plan.md Dockerfile to remove bc dependency
  13 │
  14 │ All awk replacements produce identical results to bc and have been tested.
─────┴──────────────────────────────────────────────────────────────────────────

buggerman committed:
Merge pull request #1 from buggerman/replace-bc-with-awk

─────┬───────────────────────────────────────────────────────────────…

buggerman committed:
─────┬──────────────────────────────────────────────────────────────────────────
     │ STDIN
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Fix Arrow Lake power measurement bugs
   2 │
   3 │ Fix three issues discovered with Arrow Lake CPUs on TrueNAS:
   4 │
   5 │ 1. **Prevent submission with rejected power readings**: Script now checks
   6 │    for "REJECTED" values and skips upload, matching the warning message
   7 │    behavior. Previously would show "Results will NOT be submitted" but
   8 │    still attempted upload.
   9 │
  10 │ 2. **Fix AWK division by zero**: Added check for empty/zero power data
  11 │    before division to prevent "unterminated regexp" errors when
  12 │    intel_gpu_top fails to collect power measurements.
  13 │
  14 │ 3. **Fix intel_gpu_top version check**: Changed from `--version` to `-h`
  15 │    flag for better compatibility with older versions that don't support
  16 │    --version flag, preventing "invalid option" errors.
  17 │
  18 │ These fixes improve the Arrow Lake experience on systems where power
  19 │ monitoring is unavailable, allowing users to still get valid performance
  20 │ data (FPS, speed) while properly handling missing power measurements.
─────┴──────────────────────────────────────────────────────────────────────────

buggerman committed:
Fix AWK division by zero errors in FPS and speed calculations

When FFmpeg tests fail to produce valid output (like the H.264 4K test
on some systems), the FPS and speed calculations would attempt division
by zero, causing AWK errors.

Added checks for empty/zero denominators in:
- Average FPS calculation (total_fps / fps_count)
- Average speed calculation (total_speed / speed_count)

These now default to 0.00 when no valid data is collected, matching the
behavior of the power calculation fix.

buggerman committed:
Allow submission with rejected power readings

Reverts submission blocking introduced in previous commit. Results with
rejected power readings will now be submitted to the database where they
are automatically flagged by the backend (arrow_lake_power_issue flag).

This allows Arrow Lake performance data to be collected and displayed
while appropriately marking power/efficiency metrics as unreliable.

Updated warning message to clarify:
- Results WILL be submitted but flagged
- Performance metrics (FPS, speed) remain valid and useful

buggerman committed:
Fix FPS pattern to capture single-digit frame rates

The FPS grep pattern 'fps=.[1-9][1-9].' only matched FPS values >= 10,
causing it to miss single-digit frame rates common in 4K encoding.

Changed pattern to 'fps=[0-9]+(\.[0-9]+)?' which matches:
- Single digit: 5, 7, 9
- Double digit: 15, 49, 99
- Decimal values: 5.23, 49.81

This fixes H.264 4K test showing empty FPS/speed values when encoding
runs at <10 FPS.

buggerman committed:
Capture FFmpeg progress output for FPS/speed parsing

Changed stderr redirect from /dev/null to tee ffmpeg-progress.log to
capture FFmpeg's progress output (fps=, speed=) while keeping console
clean.

This fixes missing FPS/speed data for tests that were completing
successfully but not capturing metrics due to stderr being discarded.

The -report flag creates detailed logs, but doesn't include the
progress stats we parse. Now we capture both:
- ffmpeg-*.log (detailed reports from -report flag)
- ffmpeg-progress.log (progress output with fps/speed)

buggerman committed:
Revert FFmpeg output capture - unnecessary complexity

The tee/log file capture was overkill. The H.264 4K issue was due to
a corrupt video file, not a parsing problem. The -report flag already
creates sufficient logs for FPS/speed parsing.

Reverted to simple 2>/dev/null stderr redirect.

buggerman committed:
Fix FPS pattern to handle optional space after equals

FFmpeg may output 'fps= 52.37' (with space) for slower encodes like
HEVC 4K 10-bit. Updated pattern from 'fps=[0-9]+' to 'fps= ?[0-9]+'
to handle both formats.

This fixes HEVC 4K showing 0.00 FPS despite successful encode.

Bambus3000 commented:
As a Truenas Scale user I would agree wholeheartedly, an alternative to bc would be amazing. I tried running your script, and it worked for the most part. An updated version of the most recent script with concurrency testing would be so sweet.

buggerman committed:
─────┬──────────────────────────────────────────────────────────────────────────
     │ STDIN
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Merge upstream/main: Benchmark Suite v2 + Vite/React rewrite
   2 │
   3 │ Sync fork with upstream ironicbadger/quicksync_calc. Resolved two
   4 │ conflicts to preserve the bc->awk portability work for TrueNAS:
   5 │
   6 │ - quicksync-benchmark.sh: upstream's new concurrency v2 helpers
   7 │   (run_concurrent_test, find_max_concurrency) added 4 fresh `bc`
   8 │   calls; converted them to `awk` matching the existing style so the
   9 │   script still runs on TrueNAS 25.10 without bc installed.
  10 │ - web/src/pages/about.astro: deleted upstream as part of the Astro ->
  11 │   Vite + React migration; ported the awk requirement wording into the
  12 │   new web/src/pages/about/AboutPage.tsx.
─────┴──────────────────────────────────────────────────────────────────────────

buggerman commented:
@ironicbadger Please take a look when you have time.. :)

Bambus3000 commented:
I'm currently running this script on my Truenas Scale box running an Intel Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz and an Intel Arc A310 GPU. I'm running it with --concurrency.

I would like to ask for some features:

1. Could you maybe make it not run the initial benchmark when run with the --concurrency flag, there really is no point in that right?

2. I'd like you to bump the concurrency testing limit up beyond 10x, since the A310 has plenty of power left in the tank when testing the 1080p transcodes, and I'd like to know where it actually hits the limit: 
[1/5] H.264 1080p concurrency test
Testing 10x concurrent... 2.18x <- Still plenty more to go.

[3/5] HEVC 8-bit 1080p concurrency test
Testing 10x concurrent... 1.46x <- here also, more to go.

And for whatever reason, I'm being told the A310 has no support for the experimental codecs?? Surely this is wrong.

Detecting available QSV encoders...
  VP9 QSV: not available
  AV1 QSV: not available

Then some minor stuff:
Is it possible to make the environment variable to not upload the results a flag?

buggerman commented:
> I'm currently running this script on my Truenas Scale box running an Intel Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz and an Intel Arc A310 GPU. I'm running it with --concurrency.
> 
> I would like to ask for some features:
> 
> 1. Could you maybe make it not run the initial benchmark when run with the --concurrency flag, there really is no point in that right?
> 2. I'd like you to bump the concurrency testing limit up beyond 10x, since the A310 has plenty of power left in the tank when testing the 1080p transcodes, and I'd like to know where it actually hits the limit:
>    [1/5] H.264 1080p concurrency test
>    Testing 10x concurrent... 2.18x <- Still plenty more to go.
> 
> [3/5] HEVC 8-bit 1080p concurrency test Testing 10x concurrent... 1.46x <- here also, more to go.
> 
> And for whatever reason, I'm being told the A310 has no support for the experimental codecs?? Surely this is wrong.
> 
> Detecting available QSV encoders... VP9 QSV: not available AV1 QSV: not available
> 
> Then some minor stuff: Is it possible to make the environment variable to not upload the results a flag?

Yeah seems reasonable.

buggerman committed:
Address PR feedback: probe size, concurrency cap, new CLI flags

- Fix VP9/AV1 QSV detection on Arc GPUs: bump check_qsv_encoder
  probe from 64x64:d=0.1 to 640x480:d=1. Tiny test frames silently
  fail to initialize VP9/AV1 hardware on Arc A310 / Battlemage,
  causing them to be reported as unsupported.

- Raise default concurrency cap from 10x to 20x. The existing
  speed >= 1.0x short-circuit still ends the run early on weaker
  hardware, but powerful GPUs (e.g. Arc A310 hitting 2.18x at 10x
  on H.264 1080p) can now reach their actual ceiling.

- Add --concurrency-only flag: skip single-stream benchmarks and
  run only the concurrency suite. --concurrency keeps existing
  semantics (both phases) for backwards compatibility.

- Add --max-concurrency N flag for explicit override (positive
  integer, validated). Concurrency results header is now generated
  dynamically from the cap; the API parser already accepts variable
  column counts.

- Add --no-submit flag as an alias for QUICKSYNC_NO_SUBMIT=1, so
  users can opt out without exporting an env var.

buggerman committed:
Merge pull request #2 from buggerman/feedback/probe-flags

Fix VP9/AV1 probe on Arc, raise concurrency cap, add CLI flags

buggerman committed:
docs(README): document new flags and env vars

Add an "Options" section covering --concurrency, --concurrency-only,
--max-concurrency, --no-submit, --skip-warnings, plus environment
variables (QUICKSYNC_ID, QUICKSYNC_NO_SUBMIT, QUICKSYNC_API_URL).
Includes example invocations.

buggerman commented:
@Bambus3000 Check the updated script. See the updated README for details.

Bambus3000 commented:
@buggerman Running it now, first as normal, then I'll try the new -concurrency-only flag.

VP9 and AV1 is now detected as supported. 

BTW, it's probably a limitation of Truenas, but I'm not getting any power consumption numbers.

=======================================================

       Done!

Experimental Results:

 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  hevc_4k_10bit  ribblehead_4k_hevc_10bit  44617 kb/s  63.874s  53.14   1.75x  REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  vp9_1080p      ribblehead_1080p_h264     18952 kb/s           0.00    0.00x  REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  av1_1080p      ribblehead_1080p_h264     18952 kb/s  11.730s  280.00  9.70x  REJECTED


=======================================================
All Results:

CPU                                        TEST            FILE                        BITRATE     TIME     AVG_FPS  AVG_SPEED  AVG_WATTS
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  h264_1080p_cpu  ribblehead_1080p_h264       18952 kb/s  38.452s  82.05    2.73x      N/A
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  h264_1080p      ribblehead_1080p_h264       18952 kb/s  8.699s   369.94   12.94x     REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  h264_4k         ribblehead_4k_h264          46881 kb/s  33.917s  98.24    3.29x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  hevc_8bit       ribblehead_1080p_hevc_8bit  14947 kb/s  9.863s   327.80   11.39x     REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  hevc_4k_10bit   ribblehead_4k_hevc_10bit    44617 kb/s  63.874s  53.14    1.75x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  vp9_1080p       ribblehead_1080p_h264       18952 kb/s           0.00     0.00x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  av1_1080p       ribblehead_1080p_h264       18952 kb/s  11.730s  280.00   9.70x      REJECTED

Not sure why the VP9 test failed, nor why it looks like it is transcoding the 4K file to VP9?? 
EDIT: No, I see the same result for the 4K HEVC 10bit test is shown twice.

buggerman commented:
> BTW, it's probably a limitation of Truenas, but I'm not getting any power consumption numbers.

Yes, that is very likely it.

ironicbadger commented:
Thanks for the work here. I tested this on a QuickSync node and the new flags generally behave well.

One thing I want to clarify before merging, I don’t think `--concurrency-only` should submit results by default. Concurrency results are most useful when they are tied to the same verified standard benchmark run, and the current pending-submission flow gets that linkage from the standard benchmark upload token.

So I’d prefer `--concurrency-only` to be a local measurement mode unless we add an explicit token/reuse flow later.

Could you update this PR so that:

1. `--concurrency-only` does not attempt to upload concurrency results when no standard submission token exists.
2. The README wording makes this clear. For example, instead of saying it is useful when you’ve already submitted standard results, describe it as useful for local concurrency measurement when you don’t want to rerun single-stream benchmarks.

A future enhancement could be something explicit like `--concurrency-token <token>` or a dedicated standalone concurrency submission endpoint, but I’d rather not make this mode silently submit orphaned concurrency data.

buggerman committed:
Make --concurrency-only local-only per maintainer feedback

Concurrency results need a verified standard-benchmark token to link
against. In --concurrency-only mode there is no such token, so the
prior behavior would attempt the upload and fail with a confusing
"No submission token available, skipping" message that implied a bug
rather than the intended local-only semantics.

- run_concurrency_tests: skip upload entirely when RUN_CONCURRENCY_ONLY=1
  and print a clear "results are local-only and will not be uploaded"
  message. The defensive [ -z "$SUBMISSION_TOKEN" ] guard inside
  upload_concurrency_results stays as belt-and-suspenders for the case
  where standard upload was attempted but failed.
- README: rewrite the --concurrency-only Options-table row to describe
  it as a local measurement mode rather than a follow-on submission
  mode. Update the matching example invocation comment.

A future explicit linkage flow (--concurrency-token <token> or a
standalone concurrency endpoint) is left as a follow-up; this commit
just stops silently producing orphaned concurrency submissions.

buggerman committed:
Merge pull request #3 from buggerman/feedback/concurrency-only-local

Make --concurrency-only local-only per repo owner feedback

buggerman commented:
> Thanks for the work here. I tested this on a QuickSync node and the new flags generally behave well.
>
> One thing I want to clarify before merging, I don't think `--concurrency-only` should submit results by default. Concurrency results are most useful when they are tied to the same verified standard benchmark run, and the current pending-submission flow gets that linkage from the standard benchmark upload token.
>
> So I'd prefer `--concurrency-only` to be a local measurement mode unless we add an explicit token/reuse flow later.
>
> Could you update this PR so that:
>
> 1. `--concurrency-only` does not attempt to upload concurrency results when no standard submission token exists.
> 2. The README wording makes this clear. For example, instead of saying it is useful when you've already submitted standard results, describe it as useful for local concurrency measurement when you don't want to rerun single-stream benchmarks.
>
> A future enhancement could be something explicit like `--concurrency-token <token>` or a dedicated standalone concurrency submission endpoint, but I'd rather not make this mode silently submit orphaned concurrency data.

Agreed, orphan concurrency rows shouldn't submit silently.

`--concurrency-only` now skips `upload_concurrency_results` entirely. Prints "results are local-only and will not be uploaded" instead of the old "No submission token available, skipping" which read like a bug.

README row and example reworded to local measurement mode.

Kept the `[ -z "$SUBMISSION_TOKEN" ]` guard inside `upload_concurrency_results` for the case where standard upload is attempted but fails.

`--concurrency-token` / dedicated endpoint can come as a separate PR.

Commit `390d816`.

Bambus3000 commented:
@buggerman Could you take a look to see why under the experimental test results it shows the result of the 4k 10bit hevc test as well? And it would look better if the experimental results had the CPU - TEST - FILE - BITRATE -TIME - AVG_FPS -AVG_SPEED - AVG_WATTS above them.

Experimental Results:

 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  hevc_4k_10bit  ribblehead_4k_hevc_10bit  44617 kb/s  63.872s  53.06   1.75x  REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  vp9_1080p      ribblehead_1080p_h264     18952 kb/s           0.00    0.00x  REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  av1_1080p      ribblehead_1080p_h264     18952 kb/s  11.817s  277.46  9.65x  REJECTED

=======================================================
All Results:

CPU                                        TEST            FILE                        BITRATE     TIME     AVG_FPS  AVG_SPEED  AVG_WATTS
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  h264_1080p_cpu  ribblehead_1080p_h264       18952 kb/s  39.124s  77.33    2.64x      N/A
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  h264_1080p      ribblehead_1080p_h264       18952 kb/s  8.674s   369.33   12.84x     REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  h264_4k         ribblehead_4k_h264          46881 kb/s  35.904s  93.14    3.12x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  hevc_8bit       ribblehead_1080p_hevc_8bit  14947 kb/s  9.893s   326.45   11.32x     REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  hevc_4k_10bit   ribblehead_4k_hevc_10bit    44617 kb/s  63.872s  53.06    1.75x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  vp9_1080p       ribblehead_1080p_h264       18952 kb/s           0.00     0.00x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  av1_1080p       ribblehead_1080p_h264       18952 kb/s  11.817s  277.46   9.65x      REJECTED

buggerman committed:
Fix experimental results section: header missing, last benchmark leaked

The Experimental Results block was printing tail of N+1 rows, which:

1. Leaked the final standard benchmark (hevc_4k_10bit) into the
   experimental section because tail counts from the end and the
   off-by-one was meant for a header that tail does not include.
2. Did not actually include the column header, so users had to
   match columns by position against the All Results block.

Fix prints quicksyncstats_arr index 0 (the header) explicitly,
followed by only the experimental rows. Output now shows the
correct rows under the correct header.

buggerman committed:
Merge pull request #4 from buggerman/fix/experimental-results-leak

Fix experimental results: missing header, leaked hevc_4k_10bit row

buggerman commented:
@Bambus3000 Good catch. The `tail -$((experimental_count + 1))` was an off-by-one: the `+1` looked like it was meant to include the header, but tail counts from the end so it actually leaked the last standard benchmark row (hevc_4k_10bit).

Fixed by printing the header explicitly and tailing only the experimental rows. Already in this PR (commit `95ea6f3`).

Bambus3000 commented:
@buggerman That did the trick, thanks. 

Experimental Results:

CPU                                        TEST       FILE                   BITRATE     TIME     AVG_FPS  AVG_SPEED  AVG_WATTS
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  vp9_1080p  ribblehead_1080p_h264  18952 kb/s           0.00     0.00x      REJECTED
 Intel(R) Xeon(R) Gold 6244 CPU @ 3.60GHz  av1_1080p  ribblehead_1080p_h264  18952 kb/s  11.772s  278.75   9.71x      REJECTED

Now I'd sure like to know why it fails the vp9_1080p benchmark with 0.0 FPS, the Arc Alchemist supports both decoding and encoding VP9, yet it doesn't work. Well, what can you do, I would probably never use the format anyway, just curious. 🤷‍♂️
